components/application-broker/internal/broker/provision.go

```diff
@@ -158,48 +179,81 @@ func (svc *ProvisionService) do(iID internal.InstanceID, opID internal.Operation
-	} else {
-		instanceState = internal.InstanceStateSucceeded
-		opState = internal.OperationStateSucceeded
-		opDesc = "provisioning succeeded"
-		if eventProvider {
-			err := svc.createEaOnSuccessProvision(string(appName), string(appID), string(ns), displayName, iID)
-			if err != nil {
-				instanceState = internal.InstanceStateFailed
-				opState = internal.OperationStateFailed
-				opDesc = fmt.Sprintf("provisioning failed while creating EventActivation on error: %s", err.Error())
-			}
-		}
+		svc.updateStates(iID, opID, instanceState, opState, opDesc)
+		return
+	}
+
+	if !eventProvider {
+		svc.updateStates(iID, opID, instanceState, opState, opDesc)
+		return
+	}
+
+	// create Kyma EventActivation
+	if err := svc.createEaOnSuccessProvision(appName, appID, ns, displayName, iID); err != nil {
+		instanceState = internal.InstanceStateFailed
+		opState = internal.OperationStateFailed
+		opDesc = fmt.Sprintf("provisioning failed while creating EventActivation on error: %s", err)
+		svc.updateStates(iID, opID, instanceState, opState, opDesc)
+		return
+	}
+
+	// persist Knative Subscription
+	if err := svc.persistKnativeSubscription(appName, ns); err != nil {
+		svc.log.Printf("Error persisting Knative Subscription: %v", err)
+		instanceState = internal.InstanceStateFailed
+		opState = internal.OperationStateFailed
+		opDesc = fmt.Sprintf("provisioning failed while persisting Knative Subscription for application: %s namespace: %s on error: %s", appName, ns, err)
+		svc.updateStates(iID, opID, instanceState, opState, opDesc)
+		return
+	}
+
+	// enable the namespace default Knative Broker
+	if err := svc.enableDefaultKnativeBroker(ns); err != nil {
+		instanceState = internal.InstanceStateFailed
+		opState = internal.OperationStateFailed
+		opDesc = fmt.Sprintf("provisioning failed while enabling default Knative Broker for namespace: %s on error: %s", ns, err)
+		svc.updateStates(iID, opID, instanceState, opState, opDesc)
+		return
 	}
 
+	svc.updateStates(iID, opID, instanceState, opState, opDesc)
+}
+
+func (svc *ProvisionService) updateStates(iID internal.InstanceID, opID internal.OperationID,
+	instanceState internal.InstanceState, opState internal.OperationState, opDesc string) {
+
 	if err := svc.instanceStateUpdater.UpdateState(iID, instanceState); err != nil {
-		svc.log.Errorf("Cannot update state of the stored instance [%s]: [%v]\n", iID, err)
+		svc.log.Errorf("Cannot update state of the stored instance [%s]: [%v]", iID, err)
 	}
 
 	if err := svc.operationUpdater.UpdateStateDesc(iID, opID, opState, &opDesc); err != nil {
-		svc.log.Errorf("Cannot update state for ServiceInstance [%s]: [%v]\n", iID, err)
-		return
+		svc.log.Errorf("Cannot update state for ServiceInstance [%s]: [%v]", iID, err)
 	}
 }
 
-func (svc *ProvisionService) createEaOnSuccessProvision(appName, appID, ns string, displayName string, iID internal.InstanceID) error {
+func (svc *ProvisionService) createEaOnSuccessProvision(appName internal.ApplicationName,
```

Comment:
it's not easier to pass all these arguments as 'string' type? e.g. you change namespace variable type four times (`string(ns)`)  instead of making it once during call method.

Reply:
If you ask me for my opinion I would say to not use string, because what is the point of having the typedefs then in `model.go`? Typedefs make code a bit more readable I think. But I am ok with both ways of doing it. If you prefer using string in the method signature, just let me know and I will change it :)

Reply:
I am not saying you should change the type to string in e.g. `model.go` I am saying in `createEaOnSuccessProvision` method you use only strings, e.g. `internal.Namespace` type is always converted to `string` so easier is use method as `createEaOnSuccessProvision(a,b,c string)` and `x.createEaOnSuccessProvision(string(a)...)`

Reply:
As discussed in person, we will leave it as it is

Reply:
Discussed. Can stay.